### PR TITLE
[backport-1.11] [FLINK-19748] Iterating key groups in raw keyed stream on restore fails if some key groups weren't written

### PR DESCRIPTION
**Pull request by tzulitai**
This is a backport of #13772 to `release-1.11`. Only the last 2 commits are relevant for FLINK-19748.
Please see #13772 for a detailed description.

### CODE REVIEW

**Comment by Antti-Kaikkonen** [comment]
I tried to build this from source and got the [same](https://issues.apache.org/jira/projects/FLINK/issues/FLINK-19692) error with slightly different line numbers when restoring a stateful function from a savepoint:

1)
```
git clone https://github.com/tzulitai/flink.git
cd flink
git checkout FLINK-19748-backport_1.11
mvn clean package -DskipTests
```
2)
add to flink-conf.yaml:
```
classloader.parent-first-patterns.additional: org.apache.flink.statefun;org.apache.kafka;com.google.protobuf
#optionally use rocksdb
state.backend: rocksdb
taskmanager.numberOfTaskSlots: 2
parallelism.default: 2
```

2)
Run https://github.com/Antti-Kaikkonen/FlinkStatefunCountTo1M with parallelism 2

3)
create a savepoint

4)
try to restore from the savepoint and the error is thrown in the **feedback-union -> functions** task:
```
Exception: Exception while creating StreamOperatorStateContext.
	at org.apache.flink.streaming.api.operators.StreamTaskStateInitializerImpl.streamOperatorStateContext(StreamTaskStateInitializerImpl.java:220)
	at org.apache.flink.streaming.api.operators.AbstractStreamOperator.initializeState(AbstractStreamOperator.java:248)
	at org.apache.flink.streaming.runtime.tasks.OperatorChain.initializeStateAndOpenOperators(OperatorChain.java:290)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.lambda$beforeInvoke$0(StreamTask.java:479)
	at org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor$1.runThrowing(StreamTaskActionExecutor.java:47)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.beforeInvoke(StreamTask.java:475)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:528)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:721)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:546)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: java.io.IOException: position out of bounds
	at org.apache.flink.runtime.state.StatePartitionStreamProvider.getStream(StatePartitionStreamProvider.java:58)
	at org.apache.flink.streaming.api.operators.StreamTaskStateInitializerImpl.internalTimeServiceManager(StreamTaskStateInitializerImpl.java:251)
	at org.apache.flink.streaming.api.operators.StreamTaskStateInitializerImpl.streamOperatorStateContext(StreamTaskStateInitializerImpl.java:179)
	... 9 more
Caused by: java.io.IOException: position out of bounds
	at org.apache.flink.runtime.state.memory.ByteStreamStateHandle$ByteStateHandleInputStream.seek(ByteStreamStateHandle.java:124)
	at org.apache.flink.streaming.api.operators.StreamTaskStateInitializerImpl$KeyGroupStreamIterator.next(StreamTaskStateInitializerImpl.java:458)
	at org.apache.flink.streaming.api.operators.StreamTaskStateInitializerImpl$KeyGroupStreamIterator.next(StreamTaskStateInitializerImpl.java:411)
	at org.apache.flink.streaming.api.operators.StreamTaskStateInitializerImpl.internalTimeServiceManager(StreamTaskStateInitializerImpl.java:244)
	... 10 more
```
I'm getting the same error with the default state backend and the rocksdb state backend. When I tried with rocksdb backend and heap timers I get a different error already when creating a savepoint.

**Comment by tzulitai** [comment]
@Antti-Kaikkonen thanks you for trying the branch out.

I think the exceptions you encountered are expected in the experiments you've tried out.

Can you adjust your experiments to do the following, and then report back again?:

Try out your application `FlinkStatefunCountTo1M` with a new build of StateFun that includes the changes in https://github.com/apache/flink-statefun/pull/168?

You should be able to just pull that branch, do a clean build (`mvn clean install -DskipTests`), and then change the StateFun dependency in your application to `2.3-SNAPSHOT`.

---

Let me briefly explain our release plans here to address the issue you reported, and why the above adjustment makes sense:

1. With the StateFun changes in https://github.com/apache/flink-statefun/pull/168 (and not including ANY Flink changes), we're expecting that restoring from checkpoints / savepoints should work properly now for all checkpoints / savepoints taken with a new version that includes https://github.com/apache/flink-statefun/pull/168. This would already address FLINK-19692, and we're planning to push out a StateFun hotfix release immediately to unblock you and other users that may be encountering the same issue.

2. What https://github.com/apache/flink-statefun/pull/168 doesn't yet solve, is the ability to safely restore / upgrade from a savepoint taken with StateFun versions <= 2.2.0. Enabling that requires this PR and #13761 to be fixed in Flink, release a new Flink version, and ultimately yet another follow-up StateFun hotfix releases that uses the new Flink version. That is a lengthier process, with an estimate of another 3-4 weeks.

---

TL;DR: It would be tremendously helpful if you can re-do your experiment only with a new StateFun build including https://github.com/apache/flink-statefun/pull/168 alone. Please do let me know of the results!

**Comment by Antti-Kaikkonen** [comment]
@tzulitai I tried it and restoring from a savepoint worked. As FlinkStatefunCountTo1M doesn't actually use state I also tried with my other app that uses statefun-flink-datastream and it was able to restore from a savepoint without errors. Thank you very much! 

I only tested with rocksdb state backend and rocksdb timers. The Flink version tested was 1.11.2 from https://flink.apache.org/downloads.html.

**Comment by tzulitai** [comment]
@Antti-Kaikkonen great news, thank you. We'll keep you updated on the JIRA regarding an official release candidate that includes the fixes.
